src/main/scala/is/hail/methods/Skat.scala

```diff
@@ -193,115 +132,262 @@ object Skat {
+    val R = QR.r
+    val beta = R \ (Qt * y)
+    val res = y - cov * beta
+    val sigmaSq = (res dot res) / d
+
+    val sc = keyGsWeightRdd.sparkContext
+    val resBc = sc.broadcast(res)
+    val QtBc = sc.broadcast(Qt)
+    
+    def linearTuple(x: Vector[Double], w: Double): SkatTuple = {
+      val xw = x * math.sqrt(w)
+      val sqrt_q = resBc.value dot xw
+      SkatTuple(sqrt_q * sqrt_q, xw, QtBc.value * xw)
+    }
+    
+    val maxEntries = math.max(maxSize * maxSize, maxEntriesForSmallN)
+    
+    keyGsWeightRdd
+      .map { case (key, vs) =>
+        val vsArray = vs.toArray
+        val size = vsArray.length
+        if (size <= maxSize) {
+          val skatTuples = vsArray.map((linearTuple _).tupled)
+          val (q, gramian) = if (size.toLong * n <= maxEntries) {
+            computeGramianSmallN(skatTuples)
+          } else {
+            computeGramianLargeN(skatTuples)
+          }
+          // using q / sigmaSq since Z.t * Z = gramian / sigmaSq
+          val (pval, fault) = computePval(q / sigmaSq, gramian, accuracy, iterations)
+          
+          // returning qstat = q / (2 * sigmaSq) to agree with skat R package convention
+          Row(key, size, q / (2 * sigmaSq), pval, fault)
+        } else {
+          Row(key, size, null, null, null)
+        }
+      }
+  }
 
-  /*
-  Davies algorithm takes the eigenvalues of (G * sqrt(W)).t * P_0 * G * sqrt(W)
-  We evaluate this matrix as 0.5 * (A.t * A - B.t * B) where
-  linear:   A = G * sqrt(W)              B = Q.t * G * sqrt(W)
-  logistic: A = sqrt(V) * G * sqrt(W)    B = inv(L) * X.t * V * G * sqrt(W)
-  Here X = Q * R and L * L.t = X.t * V * X
-  */
-  def runSkatPerKey(st: Array[SkatTuple], skatStatScaling: Double): SkatStat = {
+  def logisticSkat(keyGsWeightRdd: RDD[(Any, Iterable[(Vector[Double], Double)])],
+    y: DenseVector[Double], cov: DenseMatrix[Double], maxSize: Int, accuracy: Double, iterations: Int): RDD[Row] = {
+
+    val n = y.size
+    val k = cov.cols
+    val d = n - k
+
+    if (d < 1)
+      fatal(s"$n samples and $k ${ plural(k, "covariate") } including intercept implies $d degrees of freedom.")
+
+    val logRegM = new LogisticRegressionModel(cov, y).fit()
+    if (!logRegM.converged)
+      fatal("Failed to fit logistic regression null model (MLE with covariates only): " + (
+        if (logRegM.exploded)
+          s"exploded at Newton iteration ${ logRegM.nIter }"
+        else
+          "Newton iteration failed to converge"))
+
+    val mu = sigmoid(cov * logRegM.b)
+    val V = mu.map(x => x * (1 - x))
+    val VX = cov(::, *) :* V
+    val XtVX = cov.t * VX
+    XtVX.forceSymmetry()
+    var Cinv: DenseMatrix[Double] = null
+    try {
+      Cinv = inv(cholesky(XtVX))
+    } catch {
+      case _: MatrixSingularException => fatal("Singular matrix exception while computing Cholesky factor of X.t * V * X")
+      case _: NotConvergedException => fatal("Not converged exception while inverting Cholesky factor of X.t * V * X")
```

Comment:
I always prefer including the original exception, because it aids debugging, a la: 
```scala
  case e: NotConvergedException => fatal(..., e)
```